### PR TITLE
feat(debt) Add a basic snuba describe command for entities

### PR DESCRIPTION
**Pull request by fpacifici**
Adds the basic infrastructure to have `describe` commands on the snuba data model entities.
This is specifically targeted to entities. It introduces two commands:

`snuba entities list` (which returns the list of entities)

`snuba entities describe ENTITY_NAME` which show the columns + type and the relationships to other entities.

This is an example:
```
Entity groupedmessage
    Entity schema
    --------------------------------
    offset UInt64
    record_deleted UInt8
    project_id UInt64
    id UInt64
    status Nullable(UInt8)
    last_seen Nullable(DateTime)
    first_seen Nullable(DateTime)
    active_at Nullable(DateTime)
    first_release_id Nullable(UInt64)

    Relationships
    --------------------------------
        groups
        --------------------------------
        Destination: events
        Type: LEFT
            Join keys
            --------------------------------
            project_id = LEFT.project_id
            id = LEFT.group_id
```

This PR introduces an abstract description representation so we can decouple the logic that produces the content (attached to the entity, or later storage, etc.) from the formatting logic that would be different between CLI, UI and API.

Docs updated as well.

### CODE REVIEW

File: snuba/utils/describer.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    def visit_header(self, header: Optional[str]) -> None:
+        raise NotImplementedError
+
+    def visit_description(self, desc: Description) -> None:
+        raise NotImplementedError
+
+    def visit_string(self, string: str) -> None:
+        raise NotImplementedError
+
+    def visit_tuple(self, tuple: Tuple[str, str]) -> None:
```
Comment (evanh) [inline snuba/utils/describer.py:18]: This one seems out of place. It looks like if was added just to support the format of the columns in `for_schema`. I think the content should solely consist of strings or other descriptions.

**Review by evanh — COMMENTED** [review]
This definitely needs a test where you describe an entity.

File: snuba/utils/describer.py (hunk 1 above)
Comment (fpacifici) [reply snuba/utils/describer.py:18]: visit_tuple is not used for the `for_schema`. It is for everything that is conceptually a key value pair and you may want to render it in some sort of table. 
Example:
```
    Destination: events
    Type: LEFT
```

The tuple has a specific key value pair meaning, it is design for properties that have a property name and a property value. String instead have no semantics.
I can rename this to `visit_property` if it makes more sense.

Comment (evanh) [reply snuba/utils/describer.py:18]: Yeah renaming would make this more understandable.

File: snuba/cli/entities.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    Declared entities
+    """
+
+    click.echo("Declared Entities:")
+    for r in EntityKey:
+        click.echo(r.value)
+
+
+@entities.command()
+@click.argument("entity_name")
```
Comment (lynnagara) [inline snuba/cli/entities.py:52]: I think it'd be better to make this a Choice so only valid names can get passed here and the error messages are a bit nicer. Then we won't need to try/except later

**Review by lynnagara — APPROVED** [review]
This is cool.

I'd probably expect `snuba datasets list/describe` and `snuba storages list/describe` to work in the same way now. Is this what you have planned?

Also I think it'd be good to also print which dataset the entity belongs to and what storages and connected to it.

File: snuba/cli/entities.py (hunk 2 above)
Comment (fpacifici) [reply snuba/cli/entities.py]: done

File: snuba/utils/describer.py (hunk 1 above)
Comment (fpacifici) [reply snuba/utils/describer.py]: done

**Comment by fpacifici** [comment]
> I'd probably expect snuba datasets list/describe and snuba storages list/describe to work in the same way now. Is this what you have planned?

Yes we can add an equivalent for storages.

> Also I think it'd be good to also print which dataset the entity belongs to and what storages are connected to it.

I did not add that on purpose as today entities are not really strongly tied to a dataset. You load them independently from the dataset and they end up being shared. So I think we should add that when we will be enforcing this separation.

**Pull request merged**